### PR TITLE
Accurately retrieve boot environment space

### PR DESCRIPTION
**Pull request by sonicaj**
This commit does a best effort to accurately retrieve space freed if that BE is to be deleted.

Ticket: #NAS-101696

### CODE REVIEW

**Comment by sonicaj** [comment]
@william-gr there is another thing we can do here as well, we could display the full space usage for each boot environment assuming all others have been destroyed. Kindly let me know what you think about that.

**Comment by kithrup** [comment]
This may impact how PruneClone chooses — I haven’t looked at the changes yet, since I’m on a mobile device, so it may provide more useful information.

**Comment by sonicaj** [comment]
@kithrup please do have a look, on a quick peek, i found that these changes would not affect how we prune the boot environments. They are ordered from the date they are created and once they have been deleted ( if eligible ) boot pool size is checked ( regardless of what list clones reports ) and the operation of pruning continues based on that decision. However, i think we might have another bug there which is separate from this.

**Comment by kithrup** [comment]
Nope, it seems fairly straightforward and good to me!

**Comment by sonicaj** [comment]
Perfect @kithrup ;) Also had a look again and we are okay with pruneclones - no bugs : D

File: lib/Update.py [hunk 1]
```diff
@@ -464,6 +465,7 @@ def ListClones():
                     ds.properties["usedbyrefreservation"].parsed + \
                     ds.properties["usedbysnapshots"].parsed
                 origin = ds.properties["origin"].parsed
+                tdict['dataset_space'] = ds.properties['used'].parsed
```
Comment (william-gr) [inline lib/Update.py]: Thats pretty much the same as rawspace, no?

Per man zfs: `used = usedbychildren + usedbydataset + usedbyrefreservation + usedbysnapshots`

Comment (sonicaj) [reply lib/Update.py]: It changes here
```
                if '@' in origin:
                    snapshot = zfs.get_snapshot(origin)
                    tdict["rawspace"] += snapshot.properties["used"].parsed
```

File: lib/Update.py [hunk 2]
```diff
@@ -475,8 +484,65 @@ def ListClones():
...
+                # only on the space it is currently consuming as a best effort to predict.
+                # There is also "usedbysnaps" property, for that we will retrieve all snapshots of the dataset,
+                # find if any of them do not have a dataset cloned, that space will also be freed when we delete
+                # this dataset. And we will also factor in the space consumed by children.
+
+                tdict['freed_space'] = ds.properties['usedbydataset'].parsed + ds.properties['usedbychildren'].parsed
+                ds_snaps = list(ds.snapshots)
+
+                children = False
+                if ds_snaps:
```
Comment (william-gr) [inline lib/Update.py]: This `if` is superfluous

File: lib/Update.py [hunk 3]
```diff
@@ -475,8 +484,65 @@ def ListClones():
...
+                    for snap in ds_snaps:
+                        if snap.name not in datasets_origin_list:
+                            tdict['freed_space'] += snap.properties['used'].parsed
+                        else:
+                            children = True
+
+                if snapshot and not children:
+                    # This indicates the current BE is a leaf and it is safe to add the BE's origin
+                    # space to the space freed when it is deleted.
+                    tdict['freed_space'] += snapshot.properties['used'].parsed
```
Comment (william-gr) [inline lib/Update.py]: That should be `rawspace` instead of freed_space, no? since thats what used in `space`?

Comment (sonicaj) [reply lib/Update.py]: Agreed, `rawspace` is more accurate. I'll update it asap.

File: lib/Update.py [hunk 4]
```diff
@@ -475,8 +479,64 @@ def ListClones():
...
             pass
+        else:
+            if tdict['rawspace'] < 1024:
+                tdict['space'] = f'{tdict["rawspace"]}B'
+            elif 1024 <= tdict['rawspace'] < 1048576:
+                tdict['space'] = f'{tdict["rawspace"]/1024}K'
+            elif 1048576 <= tdict['rawspace'] < 1073741824:
+                tdict['space'] = f'{tdict["rawspace"]/1048576}M'
+            elif 1073741824 <= tdict['rawspace'] < 1099511627776:
+                tdict['space'] = f'{tdict["rawspace"]/1073741824}G'
```
Comment (william-gr) [inline lib/Update.py:528]: I would say anything over M should be G, we are never having over a T of boot environment, but no harm done I guess :P

**Review by william-gr — APPROVED** [review]

File: lib/Update.py (hunk 4 above)
Comment (sonicaj) [reply lib/Update.py:528]: ;)

**Pull request merged**